MetadataManager.getMetadata() throws AssertionError after storage full

While investigating why SeedVault had not run a backup since 12-12-2020 on GrapheneOS where this is implemented as the OS backup solution I found that the backup page shows a black screen and not the seed vault settings.

Device: Pixel 3a XL
GrapheneOS Build: RQ1A.210205.004.2021.02.19.15
With what I assume is the latest release of SeedVault.

ADB Logcat when navigating to GrapheneOS' backup screen (a full log taken here: [Logcat_seedvault_error.log](https://github.com/seedvault-app/seedvault/files/6032733/Logcat_seedvault_error.log)):

```02-23 16:22:10.950  1006  1043 I ActivityManager: Start proc 8292:com.stevesoltys.seedvault/u0a180 for service {com.stevesoltys.seedvault/com.stevesoltys.seedvault.transport.ConfigurableBackupTransportService}
02-23 16:22:12.738  8292  8292 E AndroidRuntime: Process: com.stevesoltys.seedvault, PID: 8292
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadata(MetadataManager.kt:43)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getBackupToken(MetadataManager.kt:195)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.App.migrateTokenFromMetadataToSettingsManager(App.kt:102)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.App.access$migrateTokenFromMetadataToSettingsManager(App.kt:37)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.App$onCreate$2.invoke(App.kt:87)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.App$onCreate$2.invoke(App.kt:37)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.AppKt.permitDiskReads(App.kt:131)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.App.onCreate(App.kt:86)
02-23 16:22:12.738  8292  8292 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadata(MetadataManager.kt:40)
02-23 16:22:12.779  1006  4808 I ActivityManager: Process com.stevesoltys.seedvault (pid 8292) has died: fg  IMPF
02-23 16:22:12.780  1006  4808 W ActivityManager: Scheduling restart of crashed service com.stevesoltys.seedvault/.transport.ConfigurableBackupTransportService in 1024000ms for connection
02-23 16:22:53.439  1006  1513 E BackupManagerService: [UserID:0] Unable to get management intent from transport: Transport com.stevesoltys.seedvault.transport.ConfigurableBackupTransport not registered
02-23 16:23:38.515  1006  4931 E BackupManagerService: [UserID:0] Unable to get management intent from transport: Transport com.stevesoltys.seedvault.transport.ConfigurableBackupTransport not registered
```

A separate crash monitored here:
```
--------- beginning of crash
02-23 16:16:24.849  3857  3857 E AndroidRuntime: FATAL EXCEPTION: main
02-23 16:16:24.849  3857  3857 E AndroidRuntime: Process: com.stevesoltys.seedvault, PID: 3857
02-23 16:16:24.849  3857  3857 E AndroidRuntime: java.lang.AssertionError: Error reading metadata from cache
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadata(MetadataManager.kt:43)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getBackupToken(MetadataManager.kt:195)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.App.migrateTokenFromMetadataToSettingsManager(App.kt:102)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.App.access$migrateTokenFromMetadataToSettingsManager(App.kt:37)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.App$onCreate$2.invoke(App.kt:87)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.App$onCreate$2.invoke(App.kt:37)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.AppKt.permitDiskReads(App.kt:131)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.App.onCreate(App.kt:86)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6715)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.app.ActivityThread.access$1300(ActivityThread.java:237)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:223)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7660)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.android.internal.os.ExecInit.main(ExecInit.java:48)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:399)
02-23 16:16:24.849  3857  3857 E AndroidRuntime: Caused by: java.io.IOException
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadata(MetadataManager.kt:40)
02-23 16:16:24.849  3857  3857 E AndroidRuntime:        ... 19 more
02-23 16:16:24.850  3857  3857 E AndroidRuntime: Error reporting crash
02-23 16:16:24.850  3857  3857 E AndroidRuntime: java.lang.RuntimeException: Bad file descriptor
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at android.os.BinderProxy.transactNative(Native Method)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at android.os.BinderProxy.transact(BinderProxy.java:550)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at android.app.IActivityManager$Stub$Proxy.handleApplicationCrash(IActivityManager.java:5208)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$KillApplicationHandler.uncaughtException(RuntimeInit.java:158)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
02-23 16:16:24.850  3857  3857 E AndroidRuntime:        at java.lang.Thread.dispatchUncaughtException(Thread.java:2203)
02-23 16:16:24.900  1006  3749 I ActivityManager: Process com.stevesoltys.seedvault (pid 3857) has died: fg  IMPF
02-23 16:16:24.901  1006  3749 W ActivityManager: Scheduling restart of crashed service com.stevesoltys.seedvault/.transport.ConfigurableBackupTransportService in 1000ms for connection
```

I'm running into the same symptom, probably the same bug. @grote, here is a backtrace you were asking for above:

```
~$ adb logcat -s MetadataManager -b all -v color
--------- beginning of main
12-08 17:15:17.453  6971  6971 E MetadataManager: Error parsing cached metadata
12-08 17:15:17.453  6971  6971 E MetadataManager: java.lang.SecurityException: org.json.JSONException: End of input at character 0 of 
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataReaderImpl.decode(MetadataReader.kt:158)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataReader$DefaultImpls.decode$default(MetadataReader.kt:31)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadataFromCache(MetadataManager.kt:250)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataManager.getMetadata(MetadataManager.kt:45)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataManager.getBackupToken(MetadataManager.kt:209)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.App.migrateTokenFromMetadataToSettingsManager(App.kt:109)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.App.access$migrateTokenFromMetadataToSettingsManager(App.kt:40)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.App$onCreate$1.invoke(App.kt:76)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.App$onCreate$1.invoke(App.kt:75)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.AppKt.permitDiskReads(App.kt:138)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.App.onCreate(App.kt:75)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1215)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6727)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.app.ActivityThread.access$1500(ActivityThread.java:248)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2054)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.os.Handler.dispatchMessage(Handler.java:106)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.os.Looper.loopOnce(Looper.java:201)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.os.Looper.loop(Looper.java:288)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at android.app.ActivityThread.main(ActivityThread.java:7841)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at java.lang.reflect.Method.invoke(Native Method)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.android.internal.os.ExecInit.main(ExecInit.java:48)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:355)
12-08 17:15:17.453  6971  6971 E MetadataManager: Caused by: org.json.JSONException: End of input at character 0 of 
12-08 17:15:17.453  6971  6971 E MetadataManager:       at org.json.JSONTokener.syntaxError(JSONTokener.java:460)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at org.json.JSONTokener.nextValue(JSONTokener.java:101)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at org.json.JSONObject.<init>(JSONObject.java:168)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at org.json.JSONObject.<init>(JSONObject.java:185)
12-08 17:15:17.453  6971  6971 E MetadataManager:       at com.stevesoltys.seedvault.metadata.MetadataReaderImpl.decode(MetadataReader.kt:89)
12-08 17:15:17.453  6971  6971 E MetadataManager:       ... 23 more
```
So looks like the `metadata.cache` file exists but is 0 bytes. I've had available disk space go to 0 as well (usable from user, not sure about system apps). Can the `metadata.cache` be rebuilt from the backup files?

--- a/app/src/main/java/com/stevesoltys/seedvault/metadata/MetadataManager.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/metadata/MetadataManager.kt
@@ -46,8 +46,12 @@ internal class MetadataManager(
                 field = try {
                     getMetadataFromCache() ?: throw IOException()
                 } catch (e: IOException) {
-                    // If this happens, it is hard to recover from this. Let's hope it never does.
-                    throw AssertionError("Error reading metadata from cache", e)
+                    // This can happen if the storage location ran out of space
+                    // or the app process got killed while writing the file.
+                    // It is hard to recover from this, so we try as best as we can here:
+                    Log.e(TAG, "ERROR getting metadata cache, creating new file ", e)
+                    // This should cause requiresInit() return true
+                    uninitializedMetadata.copy(version = (-1).toByte())
                 }
                 mLastBackupTime.postValue(field.time)
             }

--- a/app/src/main/java/com/stevesoltys/seedvault/transport/backup/BackupCoordinator.kt
+++ b/app/src/main/java/com/stevesoltys/seedvault/transport/backup/BackupCoordinator.kt
@@ -231,6 +231,7 @@ internal class BackupCoordinator(
     ): Int {
         state.cancelReason = UNKNOWN_ERROR
         if (metadataManager.requiresInit) {
+            Log.w(TAG, "Metadata requires re-init!")
             // start a new restore set to upgrade from legacy format
             // by starting a clean backup with all files using the new version
             try {

